### ISSUE TITLE
Confirm whether obsolete DisposeAsync should be reported

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#30257]
```C#
        [Fact]
        public void TestWithObsoleteDisposeAsync()
        {
            string source = @"
class C : System.IAsyncDisposable
{
    public static async System.Threading.Tasks.Task Main()
    {
        using await (var x = new C())
        {
        }
    }
    [System.Obsolete]
    public async System.Threading.Tasks.ValueTask DisposeAsync()
    {
        await System.Threading.Tasks.Task.Delay(10);
    }
}
";
            var comp = CreateCompilationWithTasksExtensions(source + s_interfaces, options: TestOptions.DebugExe);
            comp.VerifyDiagnostics();
            // PROTOTYPE(async-streams) Confirm whether this behavior is ok (currently matching behavior of obsolete Dispose in non-async using)
        }

```

**Comment by 333fred** [comment]
Note that this is the same behavior as `Dispose`, which definitely seems like bug. It would be a breaking change to fix it though...

**Comment by 333fred** [comment]
Unfortunately, it doesn't matter. People have existing code, and if they make no changes to their code or libraries then upgrading the compiler version should not break their code.

I've gone and verified that this is the same behavior as the native compiler, which reinforces this.

**Comment by 333fred** [comment]
It does seem like something that we should enforce with the warning waves stuff we're looking at, or in strict mode.

**Comment by 333fred** [comment]
So, the only reasoning for this behavior that I can think of is, if the compile-time type is does not implement `IDisposable`, we still insert a runtime check to see if it needs to be disposed. If the runtime type has an error obsolete `Dispose` method, it will still be called with no indications of an issue. I think we're likely stuck with this behavior.

**Comment by BillWagner** [comment]
@333fred @jcouv I think any resolution of this would change the compiler behavior, not the standard. I don't think I should move this issue to the dotnet/csharpstandard repo per dotnet/csharplang#173. Would you agree?

**Comment by 333fred** [comment]
I don't believe this is a spec bug, no.

**Comment by jcouv** [comment]
@BillWagner I agree this is a compiler bug (not a spec bug).

Here are the relevant LDM notes I could find to support that expectation: https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2019/LDM-2019-01-09.md#obsolete
